src/modules/ecdsa_sign_to_contract/main_impl.h

```diff
@@ -0,0 +1,230 @@
+/**********************************************************************
+ * Copyright (c) 2019 Marko Bencun, Jonas Nick                        *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef SECP256K1_MODULE_ECDSA_SIGN_TO_CONTRACT_MAIN_H
+#define SECP256K1_MODULE_ECDSA_SIGN_TO_CONTRACT_MAIN_H
+
+#include "include/secp256k1_ecdsa_sign_to_contract.h"
+
+int secp256k1_ecdsa_s2c_sign(
```

Comment:
The previous version in this PR was easier to review. I don't remember how exactly it looked like but you'd get there by doing something like this: 
* move this function to `secp256k1.c` and rename it to `secp256k1_ecdsa_sign_helper`. 
* change regular `ecdsa_sign` function to just call `ecdsa_sign_helper` function but with `s2c_data` and `s2c_opening` set to NULL.

Then we wouldn't have all this code duplication. The downside is that we wouldn't be able to fully test  `ecdsa_sign_helper` without enabling the s2c module.

Reply:
I copied it to the module with the goal of not touching `secp256k1_ecdsa_sign`. I thought that was the goal, otherwise wouldn't this nullify the purpose of an experimental module and increase the barrier to merge? 

For comparison, [the recoverable module also copied the whole function](https://github.com/bitcoin-core/secp256k1/blob/fa3301713549d118e57ebe6551d062903ddd6b63/src/modules/recovery/main_impl.h#L123) with the minimal change of extracting the `recid`.

Here, it is also a copy except for the `if(s2c_data32 != NULL) {` clauses and s2c_data32 related arg checks.

Imho it would make sense to postpone this refactoring to a future PR, and also cover the recoverable module. Since the tests for the helper and the main signing functions would overlap  >90%, the tests should also be refactored to not copy all the tests with slightly different call signature.

Reply:
> wouldn't this nullify the purpose of an experimental module and increase the barrier to merge?

No because the point is that people can use sign-to-contract only if they pass the `--enable-experimental-modules` option. That the helper doesn't change ecdsa signing if s2c_data is not set is much easier to check in review than having to diff ecdsa_sign and ecdsa_s2c_sign.

@sipa Summarizing above discussion, do you prefer copying the ecdsa code into the ecdsa_s2c module or should we make the main ecdsa function accept `s2c_data` but don't expose the argument in secp256k1.h - only in secp256k1_ecdsa_sign_to_contract?

Reply:
Just to give a quick update: I have been a bit busy the past weeks, but I still intend to move this to production on the bitbox02 (prototype already working).

I will happily refactor the current code like you suggested, then we can compare. I expect to be able to do this two weeks from now.

In the meantime, I'd still be eager to hear @sipa's opinion on how to best structure this.

Reply:
I rebased this PR, and made PR to compare both alternatives: https://github.com/bitcoin-core/secp256k1/pull/669